Sources/StreamChat/Utils/Database/NSManagedObject+Extensions.swift

```diff
@@ -163,3 +163,25 @@ extension NSManagedObjectContext {
         return objects
     }
 }
+
+// MARK: - Runtime Filtering
+
+extension Set where Element: NSManagedObject {
+    func filtered(using request: NSFetchRequest<Element>) -> [Element] {
+        guard !isEmpty else { return [] }
+        let result = NSMutableArray(array: (self as NSSet).allObjects, copyItems: false)
+        if let predicate = request.predicate {
+            result.filter(using: predicate)
+        }
+        if let sortDescriptors = request.sortDescriptors, !sortDescriptors.isEmpty {
+            result.sort(using: sortDescriptors)
+        }
+        if request.fetchLimit > 0 {
+            let removeCount = result.count - request.fetchLimit
+            if removeCount > 0 {
+                result.removeObjects(in: NSRange(location: result.count - removeCount, length: removeCount))
+            }
+        }
+        return result as? [Element] ?? []
```

Comment:
Optimisation opportunity: first sort, then run filtering in batches until we have expected amount of objects.

Reply:
This gave a boost from 105 ms to 71 ms